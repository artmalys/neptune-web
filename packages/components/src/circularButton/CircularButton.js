import React, { cloneElement } from 'react';
import Types from 'prop-types';
import classNames from 'classnames';

import './CircularButton.css';

const Type = {
  ACCENT: 'accent',
  POSITIVE: 'positive',
  NEGATIVE: 'negative',
};

const Priority = {
  PRIMARY: 'primary',
  SCONDARY: 'secondary',
};

const CircularButton = ({ className, children, disabled, icon, priority, type, ...rest }) => {
  const classes = classNames(
    'btn',
    {
      'btn-primary': type === Type.ACCENT,
      'btn-success': type === Type.POSITIVE,
      'btn-danger': type === Type.NEGATIVE,
    },
    className,
  );

  const iconEl = icon.props.size !== 24 ? cloneElement(icon, { size: 24 }) : icon;

  return (
    <label className={`np-circular-btn ${priority} ${type}`}>
      <span className="np-circular-btn__btn">
        <input
          type="button"
          aria-label={children}
          className={classes}
          disabled={disabled}
          {...rest}
        />
        {iconEl}
      </span>
      <span className="np-circular-btn__label">{children}</span>
    </label>
  );
};

CircularButton.Type = Type;
CircularButton.Priority = Priority;

CircularButton.propTypes = {
  className: Types.string,
  children: Types.string.isRequired,
  disabled: Types.bool,
  icon: Types.element.isRequired,
  priority: Types.oneOf(Object.values(Priority)),
  type: Types.oneOf(Object.values(Type)),
  onClick: Types.func.isRequired,
};

CircularButton.defaultProps = {
  className: undefined,
  priority: CircularButton.Priority.PRIMARY,
  type: CircularButton.Type.ACCENT,
  disabled: false,
};

export default CircularButton;
