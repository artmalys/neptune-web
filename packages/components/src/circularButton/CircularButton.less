.np-circular-btn {
    display: inline-flex;
    flex-direction: column;
    align-items: center;
    text-align: center;
    cursor: pointer;

    position: relative;

    input[type="button"] {
        min-height: var(--size-40);
        width: var(--size-40);
        height: var(--size-40);
        padding: 0;
        border-radius: 50%;

        margin-bottom: var(--size-8);
    }
    .tw-icon {
        position: absolute;
        top: 8px;
        pointer-events: none;
        color: white;

        transition: color .15s ease-in-out;
    }
}

.np-circular-btn__label {
    font-weight: 600;
    font-family: inherit;

    transition: color .15s ease-in-out;
}

.label-state-colours(accent,
    var(--color-control-accent),
    var(--color-control-accent-hover),
    var(--color-control-accent-active));

.label-state-colours(positive,
    var(--color-control-positive),
    var(--color-control-positive-hover),
    var(--color-control-positive-active));

.label-state-colours(negative,
    var(--color-control-negative),
    var(--color-control-negative-hover),
    var(--color-control-negative-active));

.label-state-colours(@type, @normal, @hover, @active) {
  .np-circular-btn {
        &.@{type} {
            &.secondary .tw-icon,
            .np-circular-btn__label {
                color: @normal;
            }

            &.secondary:hover .tw-icon,
            &:hover .np-circular-btn__label,
            input[type="button"]:focus ~ .np-circular-btn__label {
                color: @hover;
            }

            &.secondary:active .tw-icon,
            &:active .np-circular-btn__label,
            input[type="button"]:active ~ .np-circular-btn__label {
                color: @active;
            }
        }
    }
}
