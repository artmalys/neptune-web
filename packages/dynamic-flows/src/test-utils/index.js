import React from 'react';
import { mount } from 'enzyme';
import { Provider } from '@transferwise/components';

function mountWithProviders(component, locale = 'en-GB', messages = {}) {
  return mount(<Provider i18n={{ locale, messages }}>{component}</Provider>);
}

export { mountWithProviders as mount };
