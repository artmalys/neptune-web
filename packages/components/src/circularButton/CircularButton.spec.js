import React from 'react';
import '@testing-library/jest-dom';
import { Plus } from '@transferwise/icons';
import { render, screen } from '../test-utils';

import CircularButton from './CircularButton';

describe('CircularButton', () => {
  const props = {
    children: 'Add money',
    icon: <Plus />,
    onClick: jest.fn(),
  };

  describe('defaults', () => {
    beforeEach(() => {
      render(<CircularButton {...props} />);
    });

    it('renders the text', () => {
      screen.getByText('Add money');
    });

    it('is not disabled', () => {
      expect(screen.getByRole('button')).not.toBeDisabled();
    });

    it('renders a button of type accent and priority primary', () => {
      expect(render(<CircularButton {...props} />).container).toMatchSnapshot();
    });
  });

  describe('button attributes', () => {
    it('disables the button if disabled', () => {
      render(<CircularButton {...props} disabled />);
      expect(screen.getByRole('button')).toBeDisabled();
    });

    it('passes through custom classes if set', () => {
      render(<CircularButton {...props} className="catsarethebest" />);
      expect(screen.getByRole('label')).toHaveClass('catsarethebest');
    });
  });
});
