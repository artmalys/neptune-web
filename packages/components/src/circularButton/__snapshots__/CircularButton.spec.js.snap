// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CircularButton defaults renders a button of type accent and priority primary 1`] = `
<div>
  <label
    class="np-circular-btn primary accent undefined"
  >
    <input
      aria-label="Add money"
      class="btn np-btn btn-accent btn-priority-1"
      type="button"
    />
    <span
      aria-hidden="true"
      class="tw-icon tw-icon-plus "
      role="presentation"
    >
      <svg
        fill="currentColor"
        focusable="false"
        height="24"
        width="24"
      >
        <path
          clip-rule="evenodd"
          d="M11 22V2h2v20h-2z"
          fill-rule="evenodd"
        />
        <path
          clip-rule="evenodd"
          d="M22 13H2v-2h20v2z"
          fill-rule="evenodd"
        />
      </svg>
    </span>
    <span
      class="np-circular-btn__label"
    >
      Add money
    </span>
  </label>
</div>
`;
